pyvista/core/datasetattributes.py

```diff
@@ -711,6 +711,24 @@ def _prepare_array(
         if data is None:
             raise TypeError('``data`` cannot be None.')
 
+        # convert to numpy type if necessary
+        data = np.asanyarray(data)
+
+        if self.association == FieldAssociation.POINT:
+            array_len = self.dataset.GetNumberOfPoints()
+        elif self.association == FieldAssociation.CELL:
+            array_len = self.dataset.GetNumberOfCells()
+        else:
+            array_len = data.shape[0] if isinstance(data, np.ndarray) else 1
+
+        # Fixup input array length for scalar input
+        if not isinstance(data, np.ndarray) or np.ndim(data) == 0:
+            tmparray = np.empty(array_len)
+            tmparray.fill(data)
+            data = tmparray
+        if data.shape[0] != array_len:
+            raise ValueError(f'data length of ({data.shape[0]}) != required length ({array_len})')
+
         # attempt to reuse the existing pointer to underlying VTK data
         if isinstance(data, pyvista_ndarray):
```

Comment:
The conversion of `data` to `ndarray` makes no sense to have this if statement. In my mind, I want to remove it.

Reply:
Checks like `data.shape[0]` would fail if data isn't a numpy array. I'd leave it in.

Reply:
I'll note that none of this code was modified, the code block starting on line `732` was simply cut/pasted from its original position at the start of function to down a few lines to be executed after the initial check of the `data` shape is made.

RE: your comment, on line `733`, `isinstance(data, pyvista_ndarray)` makes sense to me here as it's the first time in the function where this check is made and it prevents an `AttributeError` from being raised for non-`pyvista_ndarray` types when accessing its `VTKObject` attribute on the next line.

However, after further review, on line `722`, i.e. `array_len = data.shape[0] if isinstance(data, np.ndarray) else 1`, the check `if isinstance(data, np.ndarray)` looks to me like it will always evaluate to `True` since the data was already cast as `ndarray`-like on line `715` with `data = np.asanyarray(data)`. I believe the intent may be to catch scalar values where `.shape` is not defined, e.g. `np.array(1).shape[0]` will raise an `IndexError`. So, perhaps line `722` should be changed from:

- `array_len = data.shape[0] if isinstance(data, np.ndarray) else 1` to
- `array_len = 1 if np.ndim(data) == 0 else data.shape[0]`

Again I'll stress that I only moved the code around and didn't change any of the lines, so what you see in this PR was already there. But, it does look like there may be a bug.

Indeed, the following code will raise an `IndexError`:
``` python
m = pv.PolyData([0, 0, 0.])
 m.point_data["foo"] = 1   # No error
m.cell_data["bar"] = 1   # No error
m["baz"] = 1  # IndexError raised
```

Interestingly, the error was raised elsewhere in `dataset`, prior to reaching the line in question (i.e. line `722` in `datasetattributes`). Here's the error traceback:
```
Traceback (most recent call last):
  File "\lib\site-packages\IPython\core\interactiveshell.py", line 3505, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-1-24ef6ab2e5f3>", line 1, in <module>
    m["baz"] = 1
  File "\pyvista\pyvista\core\dataset.py", line 1773, in __setitem__
    if scalars.shape[0] == self.n_points:
IndexError: tuple index out of range
```

Reply:
Thanks, both of you for your comments. I have to go to bed in my time zone, so let me think about it again tomorrow.

Reply:
The latest commit from yesterday includes a fix for the `IndexError` raised when setting a single scalar value for a mesh with only one point. This was kind of an aside and was not really the intent of this PR, and is a very specific and niche use case. So, to move this PR forward, should I roll back those changes, and go with the changes that were approved by @akaszynski ?